client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,406 @@
+		{
+			let endpoints = endpoints.0;
+
+			connection_notifiers_rx.close();
+			let connection_notifier_senders: Vec<_> = connection_notifiers_rx.collect().await;
+
+			connection_message.insert("msg".into(), "system.connected".into());
+
+			for (addr, verbosity) in endpoints {
+				node_map
+					.entry(id.clone())
+					.or_insert_with(Vec::new)
+					.push((verbosity, addr.clone()));
+				existing_nodes.insert(addr.clone());
+				let mut obj = serde_json::Map::new();
+				obj.insert("id".to_string(), id.into_u64().into());
+				obj.insert("payload".to_string(), connection_message.clone().into());
+				let (connection_messages, connection_notifiers) = node_args
+					.entry(addr.clone())
+					.or_insert_with(|| (Vec::new(), Vec::new()));
+				connection_messages.push(obj);
+				connection_notifiers.extend(connection_notifier_senders.clone());
+			}
+		}
 
-/// Implements `slog::Drain`.
-struct TelemetryDrain {
-	/// Sends log entries.
-	sender: std::panic::AssertUnwindSafe<mpsc::Sender<async_record::AsyncRecord>>,
-}
+		let mut node_pool: HashMap<Multiaddr, _> = existing_nodes
+			.iter()
+			.map(|addr| {
+				let (connection_messages, connection_notifiers) = node_args
+					.remove(addr)
+					.expect("there is a node for every connection message; qed");
+				let node = Node::new(
+					transport.clone(),
+					addr.clone(),
+					connection_messages,
+					connection_notifiers,
+				);
+				(addr.clone(), node)
+			})
+			.collect();
 
-/// Initializes the telemetry. See the crate root documentation for more information.
-///
-/// Please be careful to not call this function twice in the same program. The `slog` crate
-/// doesn't provide any way of knowing whether a global logger has already been registered.
-pub fn init_telemetry(config: TelemetryConfig) -> Telemetry {
-	// Build the list of telemetry endpoints.
-	let (endpoints, wasm_external_transport) = (config.endpoints.0, config.wasm_external_transport);
-
-	let (sender, receiver) = mpsc::channel(16);
-	let guard = {
-		let logger = TelemetryDrain { sender: std::panic::AssertUnwindSafe(sender) };
-		let root = slog::Logger::root(slog::Drain::fuse(logger), slog::o!());
-		slog_scope::set_global_logger(root)
-	};
-
-	let worker = match worker::TelemetryWorker::new(endpoints, wasm_external_transport) {
-		Ok(w) => Some(w),
-		Err(err) => {
-			error!(target: "telemetry", "Failed to initialize telemetry worker: {:?}", err);
-			None
+		let mut stream = receiver
+			.filter_map(|(id, verbosity, message): (Id, u8, String)| {
+				if let Some(nodes) = node_map.get(&id) {
+					future::ready(Some((verbosity, message, nodes)))
+				} else {
+					log::error!(
+						target: "telemetry",
+						"Received telemetry log for unknown id ({:?}): {}",
+						id,
+						message,
+					);
+					future::ready(None)
+				}
+			})
+			.flat_map(
+				|(verbosity, message, nodes): (u8, String, &Vec<(u8, Multiaddr)>)| {
+					let mut to_send = Vec::with_capacity(nodes.len());
```

Comment:
If you rewrite all this stuff as async code, as I already said, you would not need to do all these allocations + message cloning etc.

Please put all this into an extra function and just write everything as async code, aka not using any stream wrapper.

Reply:
You were right again :sweat: I was so focused to make this stream goes into a sink that I didn't realize the imperative-style version of this was actually easier and better. Thanks!

Done in a626cecd4